modules/collector-source/pkg/metric/diagnostics.go

```diff
@@ -0,0 +1,130 @@
+package metric
+
+import (
+	"fmt"
+	"maps"
+	"sync"
+)
+
+// Collector Metric Diagnostic IDs
+const (
+	// KubecostDiagnosticMetricID is the identifier for the metric used to determine if Kubecost metrics are being scraped.
+	KubecostDiagnosticMetricID = "kubecostMetric"
+
+	// NodesDiagnosticMetricID is the identifier for the query used to determine if the node CPU cores capacity is being scraped
+	NodesDiagnosticMetricID = "nodesCPUMetrics"
+)
+
+// diagnostic definitions mapping holds all of the diagnostic definitions that can be used for collector metrics diagnostics
+var diagnosticDefinitions map[string]*diagnosticDefinition = map[string]*diagnosticDefinition{
+	NodesDiagnosticMetricID: {
+		ID:          NodesDiagnosticMetricID,
+		MetricName:  "kube_node_status_capacity_cpu_cores",
+		Label:       "Node CPU cores capacity is being scraped",
+		Description: "Determine if the node CPU cores capacity is being scraped",
+	},
+
+	KubecostDiagnosticMetricID: {
+		ID:          KubecostDiagnosticMetricID,
+		MetricName:  "node_total_hourly_cost",
+		Label:       "Kubecost metrics for a node are being scraped",
+		Description: "Determine if kubecost metrics for a node are being scraped",
+	},
+}
+
+// diagnosticsResults stores the current state of diagnostic results
+var diagnosticsResults map[string]*diagnosticsResult = make(map[string]*diagnosticsResult)
+
+type diagnosticDefinition struct {
+	ID          string
+	MetricName  string
+	Label       string
+	Description string
+	DocLink     string
+}
+
+type diagnosticsResult struct {
+	Result map[string]any
+	Passed bool
+}
+
+type DiagnosticsModule struct {
+	lock    sync.RWMutex
+	updater Updater
+}
+
+func NewDiagnosticsModule(updater Updater) *DiagnosticsModule {
+	// Initialize diagnostics results to false to represent that no data has been collected yet
+	for id := range diagnosticDefinitions {
+		diagnosticsResults[id] = &diagnosticsResult{
+			Result: make(map[string]any),
+			Passed: false,
+		}
+	}
+
+	return &DiagnosticsModule{
+		updater: updater,
+	}
+}
+
+func (d *DiagnosticsModule) Update(updateSet *UpdateSet) {
+	if updateSet == nil {
+		return
+	}
+
+	// Create a deep copy for the async update to avoid race condition
+	updateSetCopy := &UpdateSet{
+		Timestamp: updateSet.Timestamp,
+		Updates:   make([]Update, len(updateSet.Updates)),
+	}
+	copy(updateSetCopy.Updates, updateSet.Updates)
+
+	// Run update asynchronously
+	go func() {
+		d.updater.Update(updateSetCopy)
```

Comment:
I don't think it will really matter, but instead the diagnostics module functionality should be run async rather than the lower level updater functionality. The reason for this is that it is required that updates to the repo are executed in order. If they are executed async this is no longer guaranteed.